javascript/javascript3/week2/Visual-promise/main.js

```diff
@@ -0,0 +1,32 @@
+
+function translateOneByOne() {
+    let documentChild = document.querySelectorAll('.marks > li');
+    moveElement(documentChild[0], { x: 20, y: 300 })
+        .then(() => {
+            console.log('Red circle Element has been moved');
+            moveElement(documentChild[1], { x: 400, y: 300 })
+                .then(() => {
+                    console.log('Blue circle Element has been moved');
+                    moveElement(documentChild[2], { x: 400, y: 20 })
+                        .then(() => {
+                            console.log('Green circle Element has been moved');
+                        });
+                });
+        });
+
+}
```

Comment:
This is an example of 'callback hell', it would look similar if you had used callback functions instead of promises. It works fine though, but this is the problem the that promises can solve compared to callbacks.

When using promises, you can return the promse to the next `.then` instead. So this could (should) be re-written as:
```
function translateOneByOne() {
    let documentChild = document.querySelectorAll('.marks > li');
    moveElement(documentChild[0], { x: 20, y: 300 })
        .then(() => {
            console.log('Red circle Element has been moved');
            return moveElement(documentChild[1], { x: 400, y: 300 })
        })
        .then(() => {
            console.log('Blue circle Element has been moved');
            return moveElement(documentChild[2], { x: 400, y: 20 })
        })
        .then(() => {
            console.log('Green circle Element has been moved');
        });
}
```

See how i `return` the promise each time, and it get picked up in the next `.then` - called chaining. It makes it more readable/maintainable.

Reply:
Agreed with Mads here. The 2nd version looks much better, and avoids nestedness